control/controlclient/client.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package controlclient implements the client for the Tailscale
+// control plane.
+//
+// It handles authentication, port picking, and collects the local
+// network configuration.
+package controlclient
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/empty"
+	"tailscale.com/types/netmap"
+	"tailscale.com/types/persist"
+	"tailscale.com/types/structs"
+)
+
+type Status struct {
+	_             structs.Incomparable
+	LoginFinished *empty.Message // nonempty when login finishes
+	Err           string
+	URL           string             // interactive URL to visit to finish logging in
+	Persist       *persist.Persist   // locally persisted configuration
+	NetMap        *netmap.NetworkMap // server-pushed configuration
+	Hostinfo      *tailcfg.Hostinfo  // current Hostinfo data
+
+	// The internal state machine should not be exposed outside this
+	// package, but we have some automated tests elsewhere that need to
+	// use it. Please don't use this field.
+	State State
+}
+
+// Equal reports whether s and s2 are equal.
+func (s *Status) Equal(s2 *Status) bool {
+	if s == nil && s2 == nil {
+		return true
+	}
+	return s != nil && s2 != nil &&
+		(s.LoginFinished == nil) == (s2.LoginFinished == nil) &&
+		s.Err == s2.Err &&
+		s.URL == s2.URL &&
+		reflect.DeepEqual(s.Persist, s2.Persist) &&
+		reflect.DeepEqual(s.NetMap, s2.NetMap) &&
+		reflect.DeepEqual(s.Hostinfo, s2.Hostinfo) &&
+		s.State == s2.State
+}
+
+func (s Status) String() string {
+	b, err := json.MarshalIndent(s, "", "\t")
+	if err != nil {
+		panic(err)
+	}
+	return s.State.String() + " " + string(b)
+}
+
+type Client interface {
```

Comment:
Huge interfaces are very uncommon in Go. Unlike Java, Go generally has interfaces for things you accept, rather than things you are. I guess LocalBackend accepts one of these, but I'd really rather not have this interface at all. I'd rather mock out an HTTP control server to do the right things. That seems more realistic without giving us false confidence of mock-based testing.

But we can do this for a bit. Let's rename this, though. There's no conventional pattern for these huge interface names (normally interfaces are named after the sole method they export). Maybe `ClientLike`? Then in lieu of docs (which are normally required) you can just say: 

> ClientLike is the set of methods provided by Client. It's expected that any implementation of ClientLike behave identically to Client.

Reply:
Mocking out an http server would be great at the right time, but since the control protocol is likely to change, I think it would actually create considerably more maintenance burden right now.

The best quick improvement we could do would be:
- combine registration+maprequest into a single API call (thus, only one long poll needed)
- put all the hostinfo/netinfo/endpoints updates into separate server API calls that return right away.

That would allow us to eliminate the entire gross Auto two-threaded state machine, but the Client interface at the Go level would remain about the same, not breaking ipn.Backend or this test.

After that was done safely, we could mock the server instead, which would only slightly affect the flow of this test.